src/types/lenses.ts

```diff
@@ -0,0 +1,281 @@
+  /**
+   * This method allows you to create a new lens with different shape.
+   * Keep in mind that it will not change `control.name` property.
+   *
+   * @param getter - A function that returns an object where each field is a lens.
+   *
+   * @example
+   * ```tsx
+   * function App() {
+   *   const { control, handleSubmit } = useForm<{
+   *     firstName: string;
+   *     lastName: string;
+   *   }>();
+   *
+   *   const lens = useLens({ control });
+   *
+   *   return (
+   *     <form onSubmit={handleSubmit(console.log)}>
+   *       <SharedComponent
+   *         lens={lens.reflect((l) => ({
+   *           name: l.focus('firstName'),
+   *           surname: l.focus('lastName'),
+   *         }))}
+   *       />
+   *       <input type="submit" />
+   *     </form>
+   *   );
+   * }
+   *
+   * function SharedComponent({ lens }: { lens: Lens<{ name: string; surname: string }> }) {
+   *   return (
+   *     <div>
+   *       <StringInput lens={lens.focus('name')} />
+   *       <StringInput lens={lens.focus('surname')} />
+   *     </div>
+   *   );
+   * }
+   *
+   * function StringInput({ lens }: { lens: Lens<string> }) {
+   *   return <input {...lens.interop((ctrl, name) => ctrl.register(name))} />;
+   * }
+   * ```
+   */
+  reflect: <T2 extends LensesMap>(getter: (original: Lens<T>) => T2) => Lens<UnwrapLens<T2>>;
+}
+
+export interface LensJoin<T> {
+  /**
+   * This method allows you to join two lenses into a single lens.
+   * This can be useful when you want to create a new lens from two existing lenses.
+   *
+   * @param another - The lens to join with the original lens.
+   * @param merger - A function that returns an object where each field is a lens.
+   *
+   * @example
+   * ```tsx
+   * function App() {
+   *   const { control, handleSubmit } = useForm<{
+   *     firstName: string;
+   *     lastName: string;
+   *   }>();
+   *
+   *   const lens = useLens({ control });
+   *   const firstName = lens.focus('firstName');
+   *   const lastName = lens.focus('lastName');
+   *
+   *   return (
+   *     <form onSubmit={handleSubmit(console.log)}>
+   *       <SharedComponent
+   *         lens={firstName.join(lastName, (firstNameLens, lastNameLens) => ({
+   *           name: firstNameLens,
+   *           surname: lastNameLens,
+   *         }))}
+   *       />
+   *       <input type="submit" />
+   *     </form>
+   *   );
+   * }
+   *
+   * function SharedComponent({ lens }: { lens: Lens<{ name: string; surname: string }> }) {
+   *   return (
+   *     <div>
+   *       <StringInput lens={lens.focus('name')} />
+   *       <StringInput lens={lens.focus('surname')} />
+   *     </div>
+   *   );
+   * }
+   *
+   * function StringInput({ lens }: { lens: Lens<string> }) {
+   *   return <input {...lens.interop((ctrl, name) => ctrl.register(name))} />;
+   * }
+   * ```
+   */
+  join: <T2, R extends LensesMap>(another: T2, merger: (original: Lens<T>, another: T2) => R) => Lens<UnwrapLens<R>>;
```

Comment:
There should be constraint `T2 extends Lens<any>`, but for some reason this breaks types